### PR TITLE
Cluster metrics reference

### PR DESCRIPTION
**Pull request by aarohib**
## What does this PR do?
Adds a Cluster metrics reference

### CODE REVIEW

File: kb/cluster-metrics-reference.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+**Matching Service metrics**
+
+- [poll_success_per_tl](https://github.com/temporalio/temporal/blob/6a354ee73f41a26497130f6cd0f4b97ac157f680/common/metrics/defs.go#L2823) - _need description here_
+  Example: `sum(rate(poll_success{}[5m]))`
+- [poll_timeouts](https://github.com/temporalio/temporal/blob/6a354ee73f41a26497130f6cd0f4b97ac157f680/common/metrics/defs.go#L2824) - _need description here_
+  Example: `sum(rate(poll_timeouts{}[5m]))`
+- [poll_success_sync_per_tl](https://github.com/temporalio/temporal/blob/6a354ee73f41a26497130f6cd0f4b97ac157f680/common/metrics/defs.go#L2825)
+  Example: `sum(rate(poll_success_sync{}[5m]))`
+- [asyncmatch_latency](https://github.com/temporalio/temporal/blob/77be7a96c66e6ca1ffc1f6c56d2779e0264d5dfe/common/metrics/defs.go#L2807) - measures async matched tasks from the time they are created to delivered. The larger this latency the longer tasks are sitting in the queue waiting for your workers to pick them up.
```
Comment (flossypurse) [inline kb/cluster-metrics-reference.md:91]: Already link to this source file above - don't need to do it again for each thing. The source lines could change in the future anyways.

File: kb/cluster-metrics-reference.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+```
+metrics:
+    statsd:
+        hostPort: {{ .Env.STATSD_ENDPOINT }}
+        prefix: "temporal"
+```
+
+Depending on how you deploy your services, you can also define different endpoints for each service.
+
```
Comment (flossypurse) [inline kb/cluster-metrics-reference.md:44]: This actually belongs in the Cluster configuration reference ya ?

File: kb/cluster-metrics-reference.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+- `Workflow_endtoend_latency` - measures your Workflow Execution time from start to close.
+- Temporal emits counters on completion of Workflows. These are useful in getting the overall statistics about Workflow completion. Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow completion. These metrics are also tagged with `namespace` tag.
+
+#### Cluster performance baselines
+
+Here are some baselines to consider when analyzing performance:
+
+- Frontend API latency at P95 should be below 100ms.
+- For Matching Service, - `AddWorkflowTask` ideally should be 50ms to ~100ms - `AddActivityTask` might be slightly higher if there is contention between multiple concurrent Activities for the same Workflow Execution.
+  _add more baselines here_
```
Comment (flossypurse) [inline kb/cluster-metrics-reference.md:148]: This is tapping into something next level

**Review by flossypurse — COMMENTED** [review]
I think this should be a reference in "/references"

File: kb/cluster-metrics-reference.md (hunk 2 above)
Comment (aarohib) [reply kb/cluster-metrics-reference.md]: Updated

File: kb/cluster-metrics-reference.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+Note that apart from these metrics emitted by the Cluster, you should also monitor infrastructure-specific metrics like CPU, memory, and network for all hosts that are running Temporal Cluster services.
+
+## Metrics reference
+
+Temporal emits metrics for each gRPC service request.
+These metrics are emitted with `type`, `operation`, and `namespace` tags, which provide visibility into service usage and show the request rates across services, namespaces, and even operations.
+
+- Use the `operation` tag with each metric in your query to get request rates, error rates or latencies per operation.
+- Use `service_type` with the [service role tag values](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L108) to get details for the specific service.
```
Comment (yiminc) [inline kb/cluster-metrics-reference.md:26]: The tag is `service_name` not `service_type`.

Comment (yiminc) [reply kb/cluster-metrics-reference.md:26]: And change all service_type below to service_name.

File: kb/cluster-metrics-reference.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+For example, to get the Workflow Task start to schedule ratio, use:
+
+```
+sum(rate(service_requests{service_type="history", operation="RecordWorkflowTaskStarted"}[2m])) / sum(rate(service_requests{service_type="matching", operation="AddWorkflowTask"}[2m]))
```
Comment (yiminc) [inline kb/cluster-metrics-reference.md:31]: This is not a good example to look at, like we rarely looked at this start to schedule ratio.
A better example would be service requests by operation on frontend:
sum by (operation) (rate(service_requests{service_name="frontend"}[2m]))

File: kb/cluster-metrics-reference.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+Note: All metrics queries in this article are [Prometheus queries](https://prometheus.io/docs/prometheus/latest/querying/basics/).
+
+The following list describes some metrics you can get started with.
+
+- [`service_requests`](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L2497) - shows service requests received per Task Queue.
+  Example: Service requests by operation
+  `sum(rate(service_requests{operation=\"AddWorkflowTask\"}[2m]))`
+
+- [`service_latency`](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L2502) - shows latencies for all Client request operations. Usually these are the starting point to investigate which operation is experiencing high latency issues.
+  Example: Service latency by operation for the Frontend service
```
Comment (yiminc) [inline kb/cluster-metrics-reference.md:43]: P95 service latency

File: kb/cluster-metrics-reference.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  Example:
+  - Prometheus query for getting latency by percentile:
+    `histogram_quantile($percentile, sum(rate(persistence_latency_bucket{service="$service" service_type="history"}[1m])) by (operation, le))`
+
+### Workflows and Activities metrics
+
+- To monitor task queue backlog look at the `schedule_to_start` latency metrics, for example, `temporal_workflow_task_schedule_to_start_latency_seconds_bucket` and `temporal_activity_schedule_to_start_latency_seconds_bucket` metrics.
+- For Activity Workers, use the `temporal_worker_task_slots_available` SDK metric with label `worker_type="ActivityWorker"` to help you tune your `WorkerOptions#getMaxConcurrentActivityExecutionSize` setting.
+- `Workflow_endtoend_latency` - measures your Workflow Execution time from start to close.
+- Temporal emits counters on completion of Workflows. These are useful in getting the overall statistics about Workflow completion. Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow completion. These metrics are also tagged with `namespace` tag.
```
Comment (yiminc) [inline kb/cluster-metrics-reference.md:118]: These are sdk side metrics, not from server.

File: kb/cluster-metrics-reference.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+## Cluster performance baselines
+
+Here are some baselines to consider when analyzing performance:
+
+- Frontend API latency at P95 should be below 100ms.
+- For Matching Service, - `AddWorkflowTask` ideally should be 50ms to ~100ms - `AddActivityTask` might be slightly higher if there is contention between multiple concurrent Activities for the same Workflow Execution.
```
Comment (yiminc) [inline kb/cluster-metrics-reference.md:125]: Probably should remove these.

File: kb/cluster-metrics-reference.md (hunk 5 above)
Comment (aarohib) [reply kb/cluster-metrics-reference.md:31]: updated

File: docs-src/references/cluster-metrics.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
+---
+id: cluster-metrics
+title: Cluster metrics reference
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:3]:
```suggestion
title: Temporal Cluster metrics reference
```

File: docs-src/references/cluster-metrics.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+For example, to see service requests by operation on Frontend, use:
+
+`sum by (operation) (rate(service_requests{service_name="frontend"}[2m]))`
+
+Note: All metrics queries in this article are [Prometheus queries](https://prometheus.io/docs/prometheus/latest/querying/basics/).
+
+The following list describes some metrics you can get started with.
+
+### [`service_requests`](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L2497)
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:39]:
```suggestion
### service_requests
```

I think we should not be linking headers. If it is not addressed already in the style guide, I would like to propose we add that.

Also at this point we want to generally move away from linking to specific lines in code if we can avoid it.

I think if we want to highlight where these metrics are defined in source, then maybe add a bold link at the top of the file?

File: docs-src/references/cluster-metrics.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
+---
+id: cluster-metrics
+title: Cluster metrics reference
+description: The Temporal Cluster reports a range of metrics to help operators get visibility into the Cluster’s performance and set up alerts.
+sidebar_label: Cluser metrics
+tags:
+  - reference
+---
+
+The Temporal Cluster reports a range of metrics to help operators get visibility into the Cluster’s performance and set up alerts.
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:10]:
```suggestion
A Temporal Cluster emits a range of metrics to help operators get visibility into the Cluster’s performance and to set up alerts.
```

File: docs-src/references/cluster-metrics.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+All metrics emitted by the Cluster are listed in [defs.go](https://github.com/temporalio/temporal/blob/master/common/metrics/defs.go).
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:11]:
```suggestion
All metrics emitted by the Cluster are defined in the following file: **[https://github.com/temporalio/temporal/blob/master/common/metrics/defs.go](https://github.com/temporalio/temporal/blob/master/common/metrics/defs.go)**.
```

File: docs-src/references/cluster-metrics.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+For details on setting up metrics in your Cluster configuraiton, see [Temporal Cluster](/references/configuration/##global).
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:13]:
```suggestion
For details on setting up metrics in your Cluster configuraiton, see [Temporal Cluster](/references/configuration#global).
```

File: docs-src/references/cluster-metrics.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+  `histogram_quantile($percentile, sum(rate(persistence_latency_bucket{service="$service" service_name="history"}[1m])) by (operation, le))`
+
+## Relevant SDK Metrics for Workflows and Activities
+
+- To monitor task queue backlog look at the `schedule_to_start` latency metrics, for example, `temporal_workflow_task_schedule_to_start_latency_seconds_bucket` and `temporal_activity_schedule_to_start_latency_seconds_bucket` metrics.
+- For Activity Workers, use the `temporal_worker_task_slots_available` SDK metric with label `worker_type="ActivityWorker"` to help you tune your `WorkerOptions#getMaxConcurrentActivityExecutionSize` setting.
+- `Workflow_endtoend_latency` - measures your Workflow Execution time from start to close.
+- Temporal emits counters on completion of Workflows. These are useful in getting the overall statistics about Workflow completion. Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow completion. These metrics are also tagged with `namespace` tag.
+
+## Cluster performance baselines
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:171]: I think @yiminc suggested we drop this particular bit for now.

Obviously we want to address performance baselines, but this feels wildly incomplete and thus misleading.

So my preference would be to drop this section for now - and then immediately follow up on this as a KB article.

**Review by flossypurse — CHANGES_REQUESTED** [review]

File: docs-src/references/cluster-metrics.md [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+
+Note that apart from these metrics emitted by the Cluster, you should also monitor infrastructure-specific metrics like CPU, memory, and network for all hosts that are running Temporal Cluster services.
+
+## Common metrics
+
+Temporal emits metrics for each gRPC service request.
+These metrics are emitted with `type`, `operation`, and `namespace` tags, which provide visibility into Service usage and show the request rates across Services, Namespaces, and Operations.
+
+- Use the `operation` tag in your query to get request rates, error rates or latencies per operation.
+- Use `service_name` tag with the [service role tag values](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L108) to get details for the specific service.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:27]:
```suggestion
- Use the `operation` tag in your query to get request rates, error rates, or latencies per operation.
- Use the `service_name` tag with the [service role tag values](https://github.com/temporalio/temporal/blob/bba148cf1e1642fd39fa0174423b183d5fc62d95/common/metrics/defs.go#L108) to get details for the specific service.
```

File: docs-src/references/cluster-metrics.md [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+All common tags that you can add in your query are defined in the following file
+**(https://github.com/temporalio/temporal/blob/86966c5d2f78ee74c10b6808857f1a2f64d3c134/common/metrics/defs.go#L90)**.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:30]:
```suggestion
All common tags that you can add in your query are defined in the following file:
https://github.com/temporalio/temporal/blob/86966c5d2f78ee74c10b6808857f1a2f64d3c134/common/metrics/defs.go#L90.
```

File: docs-src/references/cluster-metrics.md [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+Example: `sum(rate(poll_success{}[5m]))`
+
+### `poll_timeouts`
+
+Shows when no tasks are available for the poller within the poll timeout.
+Example: `sum(rate(poll_timeouts{}[5m]))`
+
+### `asyncmatch_latency`
+
+Measures async matched tasks from the time they are created to delivered. The larger this latency the longer tasks are sitting in the queue waiting for your workers to pick them up.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:81]:
```suggestion
Measures the time from creation to delivery for async matched Tasks.
The larger this latency, the longer Tasks are sitting in the queue waiting for your Workers to pick them up.
```

File: docs-src/references/cluster-metrics.md [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+Example: `histogram_quantile(0.95, sum(rate(asyncmatch_latency_bucket{service_name=~"matching"}[5m])) by (operation, le))`
+
+## History Service metrics
+
+History task is an internal task in Temporal that is created as part of a transaction to update Workflow state, and is processed by the Temporal History service.
+It is critical to ensure that the History task processing system is healthy.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:87]:
```suggestion
A History Task is an internal Task in Temporal that is created as part of a transaction to update Workflow state and is processed by the Temporal History service.
It is critical to ensure that the History Task processing system is healthy.
```

File: docs-src/references/cluster-metrics.md [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+The following key metrics can be used to monitor the History Service health:
+
+### `task_requests`
+
+Emit on every task process request.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:92]:
```suggestion
Emitted on every Task process request.
```

File: docs-src/references/cluster-metrics.md [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+Example: `sum(rate(task_requests{service="$service",operation=~"TransferActive.*"}[1m]))`
+
+### `task_errors`
+
+Emit on every task process error.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:97]:
```suggestion
Emitted on every Task process error.
```

File: docs-src/references/cluster-metrics.md [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+Example: `sum(rate(task_errors{operation=~"TransferActive.*"}[1m]))`
+
+### `task_attempt`
+
+Number of attempts on each task execution. A task is retried forever, and each retry increases the attempt count.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:102]:
```suggestion
Number of attempts on each Task Execution.
A Task is retried forever, and each retry increases the attempt count.
```

File: docs-src/references/cluster-metrics.md [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+Example: `histogram_quantile($percentile, sum(rate(task_attempt_bucket{service="$service",operation=~"TransferActive.*"}[1m])) by (operation, le))`
+
+### `task_latency_processing`
+
+Shows the processing latency per attempt.
+Example: `histogram_quantile($percentile, sum(rate(task_latency_processing_bucket{operation=~"TransferActive.*",service="$service", service_name="history"}[1m])) by (operation, le))`
+
+### `task_latency`
+
+Measures the task processing latency for one attempt.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:112]:
```suggestion
Measures the Task processing latency for one attempt.
```

File: docs-src/references/cluster-metrics.md [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+
+### `task_latency_queue`
+
+Measures the duration, end-to-end, from when the task should be executed (from the time it was fired), to when the task is done.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:116]:
```suggestion
Measures the duration, end-to-end, from when the Task should be executed (from the time it was fired) to when the Task is done.
```

File: docs-src/references/cluster-metrics.md [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+
+### `task_latency_load`
+
+(Available only in v1.18.0+) Measures the duration from task generation to task loading (task schedule to start latency for persistence queue).
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:120]:
```suggestion
(Available only in v1.18.0+) Measures the duration from Task generation to Task loading (Task schedule to start latency for persistence queue).
```

File: docs-src/references/cluster-metrics.md [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+
+### `task_latency_schedule`
+
+(Available only in v1.18.0+) Measures the duration from task submission (to task scheduler) to processing (task schedule to start latency for in-memory queue).
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:124]:
```suggestion
(Available only in v1.18.0+) Measures the duration from Task submission (to the Task scheduler) to processing (Task schedule to start latency for in-memory queue).
```

File: docs-src/references/cluster-metrics.md [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+
+### `queue_latency_schedule`
+
+(Available only in v1.18.0+) Measures the time it takes to schedule 100 tasks in one task channel in the host-level task scheduler.
+If there are less than 100 tasks in the task channel for 30 sesoncds, the latency will be scaled to 100 tasks upon emission.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:129]:
```suggestion
(Available only in v1.18.0+) Measures the time to schedule 100 Tasks in one Task channel in the host-level Task scheduler.
If fewer than 100 Tasks are in the Task channel for 30 seconds, the latency is scaled to 100 Tasks upon emission.
```

File: docs-src/references/cluster-metrics.md [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+Note: This is still an experimental metric and is subject to change.
+
+### `task_latency_userlatency`
+
+Shows the latency introduced because of Workflow logic. For example, if you have one Workflow scheduling many Activities or Child Workflows at the same time, it can cause a per-Workflow lock contention. The wait period for the per-Workflow lock is counted as `userlatency`.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:134]:
```suggestion
Shows the latency introduced because of Workflow logic.
For example, if you have one Workflow scheduling many Activities or Child Workflows at the same time, it can cause a per-Workflow lock contention.
The wait period for the per-Workflow lock is counted as `userlatency`.
```

File: docs-src/references/cluster-metrics.md [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+
+The `operation` tag contains details about Task type and Active vs Standby statuses, and can be used to get request rates, error rates or latencies per operation, which can help identify issues caused by database problems.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:136]:
```suggestion
The `operation` tag contains details about Task type and Active versus Standby statuses, and can be used to get request rates, error rates, or latencies per operation, which can help identify issues caused by database problems.
```

File: docs-src/references/cluster-metrics.md [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+
+## Persistence metrics
+
+Temporal Server emits metrics for every persistence database read and write.
+Some of the most important ones are:
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:141]:
```suggestion
Some of the most important ones are the following:
```

File: docs-src/references/cluster-metrics.md [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+
+### `persistence_requests`
+
+Emit on every persistence request.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:145]:
```suggestion
Emitted on every persistence request.
```

File: docs-src/references/cluster-metrics.md [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+Examples:
+
+- Prometheus query for getting total number of persistence requests by operation for the History Service:
+  `sum by (operation) (rate(persistence_requests{service="$service",service_name="history"}[1m]))`
+- Prometheus query for getting total number of persistence requests by operation for the Matching Service:
+  `sum by (operation) (rate(persistence_requests{cluster="$cluster",service_name="matching"}[5m]))`
+
+### `persistence_errors`
+
+Emit on persistence error, includes `error_type` tag. This metric is a good indicator for connection issues between Server and the Persistence store.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:155]:
```suggestion
Emitted on persistence error, includes `error_type` tag.
This metric is a good indicator for connection issues between Temporal Server and the persistence store.
```

File: docs-src/references/cluster-metrics.md [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+
+Shows the latency on persistence operations.
+Example:
+
+- Prometheus query for getting latency by percentile:
+  `histogram_quantile($percentile, sum(rate(persistence_latency_bucket{service="$service" service_name="history"}[1m])) by (operation, le))`
+
+## Relevant SDK Metrics for Workflows and Activities
+
+- To monitor task queue backlog look at the `schedule_to_start` latency metrics, for example, `temporal_workflow_task_schedule_to_start_latency_seconds_bucket` and `temporal_activity_schedule_to_start_latency_seconds_bucket` metrics.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:171]:
```suggestion
- To monitor Task Queue backlog, look at the `schedule_to_start` latency metrics; for example, the `temporal_workflow_task_schedule_to_start_latency_seconds_bucket` and `temporal_activity_schedule_to_start_latency_seconds_bucket` metrics.
```

File: docs-src/references/cluster-metrics.md [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+- For Activity Workers, use the `temporal_worker_task_slots_available` SDK metric with label `worker_type="ActivityWorker"` to help you tune your `WorkerOptions#getMaxConcurrentActivityExecutionSize` setting.
+- `Workflow_endtoend_latency` - measures your Workflow Execution time from start to close.
+- Temporal emits counters on completion of Workflows. These are useful in getting the overall statistics about Workflow completion. Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow completion. These metrics are also tagged with `namespace` tag.
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:174]:
```suggestion
- `Workflow_endtoend_latency` measures your Workflow Execution time from start to close.
- Temporal emits counters on completion of Workflows.
  These are useful in getting the overall statistics about Workflow completion.
  Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate`, and `workflow_cancel` counters for each type of Workflow completion.
  These metrics are also tagged with the `namespace` tag.
```

File: docs-src/references/configuration.md [hunk 34]
```diff
@@ -90,7 +90,7 @@ The `statsd` sections supports the following settings:
 The `prometheus` sections supports the following settings:
 
 - `framework`: The framework to use, currently supports `opentelemetry` and `tally`, default is `tally`. We plan to switch default to `opentelemetry` once its API become stable.
-- `listenAddress`: Address for prometheus to scrape metrics from.
+- `listenAddress`: Address for prometheus to scrape metrics from. The Temporal Server uses the Prometheus client API and the `listenAddress` configuration is used to listen for metrics.
 - `handlerPath`: Metrics handler path for scraper, default is `/metrics`.
```
Comment (djmagee) [inline docs-src/references/configuration.md:94]:
```suggestion
- `listenAddress`: Address for Prometheus to scrape metrics from.
  The Temporal Server uses the Prometheus client API, and the `listenAddress` configuration is used to listen for metrics.
- `handlerPath`: Metrics handler path for scraper; default is `/metrics`.
```

File: docs/references/cluster-metrics.md [hunk 35]
```diff
@@ -0,0 +1,185 @@
...
+
+Emitted on every persistence request.
+Examples:
+
+- Prometheus query for getting total number of persistence requests by operation for the History Service:
+  `sum by (operation) (rate(persistence_requests{service="$service",service_name="history"}[1m]))`
+- Prometheus query for getting total number of persistence requests by operation for the Matching Service:
+  `sum by (operation) (rate(persistence_requests{cluster="$cluster",service_name="matching"}[5m]))`
+
+### `persistence_errors`
```
Comment (yiminc) [inline docs/references/cluster-metrics.md:160]: Add one for persistence_error_with_type, it contains error_type tag.

File: docs/references/cluster-metrics.md [hunk 36]
```diff
@@ -0,0 +1,185 @@
...
+### `client_errors`
+
+An indicator for connection issues between different Server roles.
+Example: Client errors
+`sum(rate(client_errors{service_name="frontend",service_role="history"}[5m]))`
+
+In addition to these, you can define some service-specific metrics to get performance details for each service.
+Start with the following list, and use [defs.go](https://github.com/temporalio/temporal/blob/master/common/metrics/defs.go) to define additional metrics as required.
+
+## Matching Service metrics
```
Comment (yiminc) [inline docs/references/cluster-metrics.md:70]: One interesting metric from matching is no_poller_tasks, it is a counter metric that is emitted whenever a task is added to a task queue that has no poller. This usually indicate wrong task queue is used either by the starter or by the worker.

File: docs-src/references/cluster-metrics.md [hunk 37]
```diff
@@ -0,0 +1,196 @@
...
+
+### `persistence_latency`
+
+Shows the latency on persistence operations.
+Example:
+
+- Prometheus query for getting latency by percentile:
+  `histogram_quantile($percentile, sum(rate(persistence_latency_bucket{service="$service" service_name="history"}[1m])) by (operation, le))`
+
+## Relevant SDK Metrics for Workflows and Activities
```
Comment (djmagee) [inline docs-src/references/cluster-metrics.md:188]:
```suggestion
## Relevant SDK metrics for Workflows and Activities
```

**Review by djmagee — APPROVED** [review]
Exactly one nit [that I likely overlooked during my previous review].

File: docs-src/references/cluster-metrics.md [hunk 38]
```diff
@@ -0,0 +1,196 @@
...
+
+### `persistence_latency`
+
+Shows the latency on persistence operations.
+Example:
+
+- Prometheus query for getting latency by percentile:
+  `histogram_quantile($percentile, sum(rate(persistence_latency_bucket{service="$service" service_name="history"}[1m])) by (operation, le))`
+
+## Relevant SDK metrics for Workflows and Activities
```
Comment (flossypurse) [inline docs-src/references/cluster-metrics.md:188]: This section is out of scope in this reference I think -

**Review by flossypurse — APPROVED** [review]
